plugins/auth-backend/src/providers/microsoftaks/provider.ts

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import express from 'express';
+import passport from 'passport';
+import { Strategy as OAuth2Strategy } from 'passport-oauth2';
+import {
+  encodeState,
+  OAuthAdapter,
+  OAuthEnvironmentHandler,
+  OAuthHandlers,
+  OAuthProviderOptions,
+  OAuthRefreshRequest,
+  OAuthResult,
+  OAuthStartRequest,
+} from '../../lib/oauth';
+import {
+  executeFetchUserProfileStrategy,
+  executeFrameHandlerStrategy,
+  executeRedirectStrategy,
+  executeRefreshTokenStrategy,
+  PassportDoneCallback,
+} from '../../lib/passport';
+import { OAuthStartResponse } from '../types';
+import { createAuthProviderIntegration } from '../createAuthProviderIntegration';
+
+type PrivateInfo = {
+  refreshToken: string;
+};
+
+export type MicrosoftAksAuthProviderOptions = OAuthProviderOptions & {
+  tenantId: string;
+};
+
+export class MicrosoftAksProvider implements OAuthHandlers {
```

Comment:
This is more or less the same as the existing `microsoft` provider right, except the default scope? In that case we should be able to rely on the existing provider and can skip this addition.

Reply:
When Microsoft redirects a user back to Backstage with an authorization code and the `MicrosoftAuthProvider`'s `handler()` method is called, then the `executeFrameHandlerStrategy` helper [calls the `MicrosoftStrategy`'s `authenticate` method](https://github.com/backstage/backstage/blob/f4e4f04152818b7ce684b62cd42b07514158bcf9/plugins/auth-backend/src/lib/passport/PassportStrategyHelper.ts#L128).  Well, `MicrosoftStrategy` [inherits from `OAuth2Strategy`](https://github.com/seanfisher/passport-microsoft/blob/29c5a91164a980f7b9845b51cc96827b304e20a0/lib/strategy.js#L58-L62), so after exchanging the authorization code for an access token, the parent implementation [attempts to fetch a user profile](https://github.com/jaredhanson/passport-oauth2/blob/5fcd4c5238a3785794c8c8d0b4a1ab94f1f67bc1/lib/strategy.js#L180).  But here `MicrosoftStrategy` has overloaded the implementation to [necessarily make a Microsoft Graph API call](https://github.com/seanfisher/passport-microsoft/blob/29c5a91164a980f7b9845b51cc96827b304e20a0/lib/strategy.js#L94).  When you remove `User.Read` from the default scopes, the access token used to make this call lacks the necessary privileges and the call fails, causing the overall flow [started by `getAccessToken()`](https://github.com/backstage/backstage/pull/15614/files#diff-63c2a4b51f05b6c09359c46da3635ed25b4d40ffe2d74e2a2d61259d392b61d9R36) (in the Kubernetes plugin) to fail.

The way we got around this was by adding `skipUserProfile: true` [here](https://github.com/backstage/backstage/pull/15614/files#diff-7c54a71e6e5c43d18f420d965f4ddad4127ee6da4d26e9b25e139aee001495b0R66), which is the other main difference.

It might be possible to parameterize `MicrosoftAuthProvider` so that we don't need a whole other class and instead we finesse the [default factories](https://github.com/backstage/backstage/pull/15614/files#diff-2c4c9015b1e5166def5c2014243e2bab9f9ac7745dd9fc4a094823f5bbb0001eR82) (or somewhere else) to ensure that redirects through `/api/auth/microsoftaks/handler/frame` do not trigger a user profile fetch from Microsoft Graph.

Reply:
Alright, makes sense. Yep I do think we want to find a solution that doesn't end up adding an additional auth provider implementation in the backend. It could actually be that it makes sense to add `azureAuthApiRef` or something like that in the frontend though, and install the microsoft provider with a separate configuration by default. More on that in the other comment though.